igvc_navigation/src/mapper/octomapper.cpp

```diff
@@ -0,0 +1,215 @@
+  for (octomap::OcTree::iterator it = pc_map_pair.octree->begin(), end = pc_map_pair.octree->end(); it != end; ++it)
+  {
+    // If this is a leaf at max depth, then only update that node
+    if (it.getDepth() == pc_map_pair.octree->getTreeDepth())
+    {
+      auto [x, y] = toMapCoordinates(it.getX(), it.getY());
+      if (x < map_options_.lengthGrid() && y < map_options_.widthGrid())
+      {
+        odds_sum[x][y] += it->getLogOdds();
+      }
+      else
+      {
+        ROS_ERROR_STREAM_THROTTLE_NAMED(10, "Point Outside", "Point outside!");
+      }
+    }
+    else
+    {
+      // This isn't a leaf at max depth. Time to iterate
+      int grid_num = 1u << (pc_map_pair.octree->getTreeDepth() - it.getDepth());
+      auto [x, y] = toMapCoordinates(it.getX(), it.getY());
+      for (int dx = 0; dx < grid_num; dx++)
+      {
+        for (int dy = 0; dy < grid_num; dy++)
+        {
+          if (x < map_options_.lengthGrid() && y < map_options_.widthGrid())
+          {
+            odds_sum[x + dx][y + dy] += it->getLogOdds();
+          }
+          else
+          {
+            ROS_ERROR_STREAM_THROTTLE_NAMED(10, "Point Outside", "Point outside!");
+          }
+        }
+      }
+    }
+  }
+
+  // Transfer from log odds to normal probability
+  for (int i = 0; i < map_options_.lengthGrid(); i++)
+  {
+    for (int j = 0; j < map_options_.widthGrid(); j++)
+    {
+      pc_map_pair.map->at<uchar>(i, j) = toCharProb(fromLogOdds(odds_sum[i][j]));
+    }
+  }
+}
+
+void Octomapper::create_map(pc_map_pair &pair) const
+{
+  int length = static_cast<int>(map_options_.lengthGrid());
+  int width = static_cast<int>(map_options_.widthGrid());
+  pair.map = boost::make_shared<cv::Mat>(length, width, map_encoding_, 127);
+}
+
+void Octomapper::insertScan(const tf::Point &sensor_pos, struct pc_map_pair &pair, const PointCloud &pc,
+                            ProbabilityModel model, double range) const
+{
+  double old_prob_hit = pair.octree->getProbHit();
+  double old_prob_miss = pair.octree->getProbMiss();
+
+  pair.octree->setProbHit(model.prob_hit);
+  pair.octree->setProbMiss(model.prob_miss);
+
+  octomap::point3d sensor = octomap::pointTfToOctomap(sensor_pos);
+  sensor.z() = 0;
+  octomap::Pointcloud octo_cloud;
+  PCL_to_Octomap(pc, octo_cloud);
+  pair.octree->insertPointCloud(octo_cloud, sensor, range, false, false);  // No discretize, no lazy eval
+
+  pair.octree->setProbHit(old_prob_hit);
+  pair.octree->setProbMiss(old_prob_miss);
+}
+
+void Octomapper::insertRays(const tf::Point &sensor_pos, struct pc_map_pair &pair, const PointCloud &pc, bool occupied,
+                            ProbabilityModel model) const
+{
+  double old_prob_hit = pair.octree->getProbHit();
+  double old_prob_miss = pair.octree->getProbMiss();
+
+  pair.octree->setProbHit(model.prob_hit);
+  pair.octree->setProbMiss(model.prob_miss);
+
+  octomap::point3d sensor = octomap::pointTfToOctomap(sensor_pos);
+  sensor.z() = 0;
```

Comment:
What's the justification for setting the sensor height to 0 when inserting a ray into the map?

Reply:
we're not actually mapping in 3d. All points are projected to 2d first before inserted into the octree, effectively treating it as a quadtree.

Reply:
Yes but if you are setting the sensor height to zero then wouldn't projections now be offset in the z-axis by the sensor height? For example: 

1. You project the image free space to the ground plane (this projection is now at z = 0 without ground plane filtering)
2. You set the sensor height to z = 0
3. Now the ground plane projection relative to the sensor is offset by -sensor_height because of step 2

Reply:
Not true. Note that this happens in `octomapper.cpp`, not `mapper.cpp`,which means this happens after projections are done and doesn't change the image projections in any way.. Also, sensor_pos is only used in `insertScan` and `insertRay`. for `insertPoint`, the points are inserted directly into the map and no "raycasting" happens.

Reply:
O.k. This should not then be a problem